src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Roslyn.Utilities;
+using LSP = Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.Handler.SemanticTokens
+{
+    /// <summary>
+    /// Computes the semantic tokens for a given range.
+    /// </summary>
+    /// <remarks>
+    /// When a user opens a file, it can be beneficial to only compute the semantic tokens for the visible range
+    /// for faster UI rendering.
+    /// The range handler is only invoked when a file is opened. When the first whole document request completes
+    /// via <see cref="SemanticTokensHandler"/>, the range handler is not invoked again for the rest of the session.
+    /// </remarks>
+    [ExportLspMethod(LSP.SemanticTokensMethods.TextDocumentSemanticTokensRangeName), Shared]
+    internal class SemanticTokensRangeHandler : AbstractRequestHandler<LSP.SemanticTokensRangeParams, LSP.SemanticTokens>
+    {
+        private readonly SemanticTokensCache _tokensCache;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SemanticTokensRangeHandler(
+            ILspSolutionProvider solutionProvider,
+            SemanticTokensCache tokensCache) : base(solutionProvider)
+        {
+            _tokensCache = tokensCache;
+        }
+
+        public override async Task<LSP.SemanticTokens> HandleRequestAsync(
+            LSP.SemanticTokensRangeParams request,
+            LSP.ClientCapabilities clientCapabilities,
+            string? clientName,
+            CancellationToken cancellationToken)
+        {
+            Contract.ThrowIfNull(request.TextDocument);
+            var requestId = _tokensCache.GetNextResultId();
+
+            // The results from the range handler should not be involved in caching since we don't
```

Comment:
then why do we have a _tokensCache? :)

Reply:
Oops, I updated the comment to make it clearer that the range handler _uses_ the cache, but doesn't update it.